File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/blocks/analysis/Contract.kt
// ── blocks/analysis/contract — the tiered relay's transport (11) ───────────
// A deep tier publishes conclusions; a fast tier recalls them. Both halves are
// ORDINARY VERBS with ordinary Command cases — the tiering lives in which
// registration list a tier is wired with (app/wire's FAST_TIER / DEEP_TIER), not in
// a second mechanic.
//
// PACKAGE NOTE (G12, in Kotlin): this file sits in blocks/analysis/ but declares package
// adr.contract, because Kotlin requires every variant of a sealed hierarchy to live
// in one package. Gate check C2 compensates.
//
// HARD CONSTRAINT (G1 + 11.3): no case declares an Actor, an Authority or a
// Signature — so A RECALLED CONCLUSION CANNOT CARRY AUTHORITY. `Recall` itself
// declares only text and publishedAt. Recall confers no permission: it is
// unrepresentable, not merely unused, which is the same bar G1 cleared.

package adr.contract

import adr.spine.pure.CommandId
import adr.spine.pure.EffectClass
import adr.spine.pure.Recall
import adr.spine.pure.Signature
import adr.spine.pure.Timestamp
import adr.spine.pure.ToolName

/**
 * A sealed CLASS extending the sealed CLASS ToolResult: `tool` is passed up the chain,
 * so every variant carries it by construction rather than by re-implementing it.
 */
sealed class AnalysisResult(override val tool: ToolName) : ToolResult(tool) {
    /**
     * The FAST tier's read. It returns the `Recalled` snapshot the consumer already
     * staged and bounded — it never reaches the relay itself, so a tool body stays
     * pure and a re-fold resolves the same snapshot from committed bytes alone.
     */
    data class RecallAnalysis(
        override val tool: ToolName,
        val recall: Recall,
    ) : AnalysisResult(tool)

    /** The DEEP tier's write. Its conclusion leaves as an effect descriptor, not a call. */
    data class PublishAnalysis(
        override val tool: ToolName,
        val text: String,
    ) : AnalysisResult(tool)
}

/**
 * A sealed CLASS extending the sealed CLASS Command: tool/sig/id pass up the chain and
 * every variant carries authorship, permission and identity by construction (G12).
 */
sealed class AnalysisCommand(
    override val tool: ToolName,
    override val sig: Signature,
    override val id: CommandId,
) : Command(tool, sig, id) {
    data class RecallAnalysis(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
        val recall: Recall,
    ) : AnalysisCommand(tool, sig, id)

    data class PublishAnalysis(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
        val text: String,
    ) : AnalysisCommand(tool, sig, id)
}

sealed class AnalysisEffect(
    override val at: Timestamp,
    effectClass: EffectClass,
) : Effect(at, effectClass) {
    /**
     * The deep tier's own signed act. Emitted ONLY by the publish arm, so recalled
     * content cannot even reach the relay — let alone an irreversible effect.
     *
     * Being an effect descriptor (14.2) buys the whole recovery story for free: it is
     * replay-stubbed in REPLAY mode and idempotency-keyed by EffectKey in RECOVERY,
     * through machinery that already ships.
     *
     * ROUTINE, and the classification is the REGISTRY'S rather than a judgement: the
     * verb that emits it (`publishAnalysis`) is registered Reversible, and an effect
     * class stricter than the verb that earns it would make the reference refuse its
     * own publish on every run. A tier write is re-drivable and dedupes on EffectKey
     * in RECOVERY. Promoting it means promoting the VERB first, with the
     * `requestedBy` lens 14.3 requires.
     */
    data class PublishConclusion(
        override val at: Timestamp,
        val text: String,
    ) : AnalysisEffect(at, EffectClass.Routine)
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/analysis/contract.ts
// ── blocks/analysis/contract — the TIER RELAY's transport (11.2, 11.3) ─────
// Two verbs, one per tier, and they are ordinary verbs: same declarations, same
// signer, same blast radius as `setPriority`. Tiering is not a second mechanic.
//
//   recallAnalysis   FAST tier. Reads the snapshot the consumer already staged.
//   publishAnalysis  DEEP tier. Emits the one effect that writes to the relay.
//
// WHY A RECALLED CONCLUSION CANNOT REACH AN IRREVERSIBLE EFFECT (10.2, 11.3).
// Three facts, all carried by machinery that already ships:
//
//   1. It cannot carry an Authority. `Recall` declares `text` and `publishedAt`
//      and nothing else, and THIS FILE is under check C4 — a block's contract
//      may not name Actor, Authority or Signature. Unrepresentable, not unused.
//   2. It cannot become a pending request. An irreversible verb's `requestedBy`
//      returns an Authority read out of committed State; the analysis slice
//      stores text with no Authority field to return, and C4 denies importing
//      `authority` inside a block's tools, so one cannot be minted either.
//   3. The recall arm emits NO effect at all. `PublishConclusion` is emitted
//      only by `publishAnalysis` — the deep tier's own signed act.
//
// So an irreversible act SUGGESTED by recalled text still has to pass the
// boundary gate on its own merits, and with no pending request from a different
// principal it is Refused. Recall confers no authority; it buys the model
// nothing. Both halves of that are tested.

import type { CommandBase } from "@adr/spine/pure/command";
import type { EffectBase } from "@adr/spine/pure/effect";
import type { Recall } from "@adr/spine/pure/staged";
import type { ToolResultBase } from "@adr/spine/pure/tool-result";
import { claims } from "@adr/spine/pure/tool-result";

// ── ToolResult cases ────────────────────────────────────────────────────────
/** Carries the whole sealed `Recall` — text AND variant. That is what makes a
 *  re-fold resolve the same snapshot AND the same branch from committed bytes
 *  alone, with no second capture log and no re-query. */
export interface RecallAnalysisResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "recallAnalysis";
  readonly recall: Recall;
}

export interface PublishAnalysisResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "publishAnalysis";
  readonly text: string;
}

export type AnalysisResult = RecallAnalysisResult | PublishAnalysisResult;

// ── Command cases ───────────────────────────────────────────────────────────
export interface RecallAnalysisCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "recallAnalysis";
  readonly recall: Recall;
}

export interface PublishAnalysisCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "publishAnalysis";
  readonly text: string;
}

export type AnalysisCommand = RecallAnalysisCommand | PublishAnalysisCommand;

// ── Effect cases ────────────────────────────────────────────────────────────
/** The deep tier's write to the relay, as an EFFECT DESCRIPTOR — so it is
 *  replay-stubbed for free and idempotency-keyed by `EffectKey` in RECOVERY,
 *  through machinery that already ships. Publishing is not a side channel. */
export interface PublishConclusion extends EffectBase {
  readonly kind: "PublishConclusion";
  /** ROUTINE, and the classification is the REGISTRY'S rather than a judgement:
   *  the verb that emits it (`publishAnalysis`) is registered `Reversible`
   *  (blocks/analysis/tools.ts), and an effect class stricter than the verb that
   *  earns it would make the reference refuse its own publish on every run. A
   *  tier write is re-drivable and dedupes on `EffectKey` in RECOVERY. Promoting
   *  it means promoting the VERB first, with the `requestedBy` lens 14.3
   *  requires. */
  readonly effectClass: "Routine";
  readonly text: string;
}

export type AnalysisEffect = PublishConclusion;

/** WHICH RESULTS THIS BLOCK'S ARM FOLDS, derived rather than written. The table
 *  is a mapped type over the union above (`spine/pure/tool-result`), so a case
 *  added there and a case claimed here are ONE edit: omit it and the property is
 *  missing, name a tool the union does not declare and the property is excess.
 *  The under-claiming predicate this used to be — green build, stale clause,
 *  fall-through at run time — is not writable in this form. */
export const isAnalysisResult = claims<AnalysisResult>({
  recallAnalysis: true,
  publishAnalysis: true,
});
